File: tests/verify_install.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""
verify_install.py - ZEN_AI_RAG Installation Verification

Checks that all required dependencies are properly installed.
Run this after installation to verify everything is ready.

Usage:
    python verify_install.py
"""

import sys
import importlib
from pathlib import Path

# Color codes for terminal output
GREEN = "\033[92m"
RED = "\033[91m"
YELLOW = "\033[93m"
CYAN = "\033[96m"
RESET = "\033[0m"
BOLD = "\033[1m"


def check_package(import_name: str, display_name: str, optional: bool = False) -> bool:
    """
    Check if a package is installed.

    Args:
        import_name: Name to use in import statement
        display_name: Human-friendly display name
        optional: If True, missing is a warning not error

    Returns:
        True if package found, False otherwise
    """
    try:
        __import__(import_name)
        status = f"{GREEN}✓{RESET}"
        category = "REQUIRED" if not optional else "OPTIONAL"
        # [X-Ray auto-fix] print(f"  {status} {display_name:45} [{category}]")
        return True
    except ImportError:
        status = f"{RED}✗{RESET}"
        category = "WARNING" if optional else "ERROR"
        # [X-Ray auto-fix] print(f"  {status} {display_name:45} [{category}]")
        return False


def check_file(filepath: str, display_name: str, optional: bool = False) -> bool:
    """Check if a file/directory exists."""
    path = Path(filepath)
    exists = path.exists()
    status = f"{GREEN}✓{RESET}" if exists else f"{RED}✗{RESET}"
    category = "WARNING" if optional and not exists else "REQUIRED" if not exists else "OK"
    # [X-Ray auto-fix] print(f"  {status} {display_name:45} [{category}]")
    return exists


def _do_do_main_setup_setup():
    """Helper: setup phase for _do_main_setup."""

    print("\n" + "=" * 85)
    # [X-Ray auto-fix] print(f"{BOLD}{CYAN}ZEN_AI_RAG Installation Verification{RESET}")
    print("=" * 85 + "\n")

    results = {
        "required": {"passed": 0, "failed": 0},
        "optional": {"passed": 0, "failed": 0},
        "files": {"passed": 0, "failed": 0},
    }

    # Core dependencies (required)
    # [X-Ray auto-fix] print(f"{BOLD}Core Dependencies (REQUIRED):{RESET}")
    core_packages = [
        ("nicegui", "NiceGUI - Web UI Framework"),
        ("uvicorn", "Uvicorn - ASGI Server"),
        ("requests", "Requests - HTTP Client"),
        ("httpx", "HTTPX - Async HTTP"),
        ("bs4", "BeautifulSoup4 - HTML Parsing"),
    ]
    for import_name, display_name in core_packages:
        if check_package(import_name, display_name):
            results["required"]["passed"] += 1
        else:
            results["required"]["failed"] += 1
    print()

    # LLM & RAG (required)
    # [X-Ray auto-fix] print(f"{BOLD}LLM & RAG Integration (REQUIRED):{RESET}")
    rag_packages = [
        ("qdrant_client", "Qdrant Client - Vector Database"),
        ("sentence_transformers", "Sentence Transformers - Embeddings"),
        ("rank_bm25", "BM25 - Ranking Algorithm"),
    ]
    for import_name, display_name in rag_packages:
        if check_package(import_name, display_name):
            results["required"]["passed"] += 1
        else:
            results["required"]["failed"] += 1
    print()

    return display_name, import_name, results


def _do_main_setup():
    """Helper: setup phase for main."""
    display_name, import_name, results = _do_do_main_setup_setup()
    # Document Processing (required)
    # [X-Ray auto-fix] print(f"{BOLD}Document Processing (REQUIRED):{RESET}")
    doc_packages = [
        ("PyPDF2", "PyPDF2 - PDF Processing"),
        ("fitz", "PyMuPDF - PDF/Image Processing"),
        ("pypdf", "PyPDF - Modern PDF Library"),
    ]
    for import_name, display_name in doc_packages:
        if check_package(import_name, display_name):
            results["required"]["passed"] += 1
        else:
            results["required"]["failed"] += 1
    print()

    # Audio (required for voice)
    # [X-Ray auto-fix] print(f"{BOLD}Audio Processing (REQUIRED for Voice):{RESET}")
    audio_packages = [
        ("faster_whisper", "Faster Whisper - STT"),
        ("piper", "Piper TTS - Text-to-Speech"),
        ("sounddevice", "SoundDevice - Audio I/O"),
    ]
    for import_name, display_name in audio_packages:
        if check_package(import_name, display_name):
            results["required"]["passed"] += 1
        else:
            results["required"]["failed"] += 1
    print()

    return display_name, import_name, results


def _do_main_init():
    """Helper: setup phase for main."""

    display_name, import_name, results = _do_main_setup()
    # ML & Utilities (required)
    # [X-Ray auto-fix] print(f"{BOLD}ML & Utilities (REQUIRED):{RESET}")
    util_packages = [
        ("numpy", "NumPy - Numerical Computing"),
        ("scipy", "SciPy - Scientific Computing"),
        ("pydantic", "Pydantic - Data Validation"),
        ("dotenv", "Python-dotenv - Config Management"),
    ]
    for import_name, display_name in util_packages:
        if check_package(import_name, display_name):
            results["required"]["passed"] += 1
        else:
            results["required"]["failed"] += 1
    print()

    # Testing (optional)
    # [X-Ray auto-fix] print(f"{BOLD}Testing Tools (OPTIONAL):{RESET}")
    test_packages = [
        ("pytest", "Pytest - Testing Framework"),
        ("pytest_asyncio", "Pytest AsyncIO - Async Testing"),
    ]
    for import_name, display_name in test_packages:
        if check_package(import_name, display_name, optional=True):
            results["optional"]["passed"] += 1
        else:
            results["optional"]["failed"] += 1
    print()

    # Vision (optional)
    # [X-Ray auto-fix] print(f"{BOLD}Vision Tools (OPTIONAL):{RESET}")
    vision_packages = [
        ("cv2", "OpenCV - Computer Vision"),
        ("PIL", "Pillow - Image Processing"),
    ]
    for import_name, display_name in vision_packages:
        if check_package(import_name, display_name, optional=True):
            results["optional"]["passed"] += 1
        else:
            results["optional"]["failed"] += 1
    print()

    return display_name, results


def main():
    """Run verification checks."""
    display_name, results = _do_main_init()
    # Directory structure
    # [X-Ray auto-fix] print(f"{BOLD}Directory Structure:{RESET}")
    dirs = [
        ("zena_mode", "Core application module"),
        ("ui", "UI components"),
        ("local_llm", "Local LLM management"),
        ("models", "AI model storage directory"),
        ("qdrant_storage", "Vector database storage"),
    ]
    for dirname, display_name in dirs:
        if check_file(dirname, display_name, optional=True):
            results["files"]["passed"] += 1
        else:
            results["files"]["failed"] += 1
    print()

    # Summary
    print("=" * 85)
    # [X-Ray auto-fix] print(f"{BOLD}Summary:{RESET}")
    print("=" * 85)

    required_ok = results["required"]["failed"] == 0
    results["optional"]["failed"] == 0

    print(
        f"\n  Required Packages: {GREEN}{results['required']['passed']}{RESET} passed, {RED}{results['required']['failed']}{RESET} failed"
    )
    print(
        f"  Optional Packages: {GREEN}{results['optional']['passed']}{RESET} passed, {YELLOW}{results['optional']['failed']}{RESET} warnings"
    )
    print(
        f"  Directory Structure: {GREEN}{results['files']['passed']}{RESET} OK, {YELLOW}{results['files']['failed']}{RESET} missing (optional)"
    )

    if required_ok:
        # [X-Ray auto-fix] print(f"\n{GREEN}{BOLD}✓ Installation Verified - Ready to Use!{RESET}")
        # [X-Ray auto-fix] print(f"\nNext: Run 'python zena.py' to start the application")
        print("=" * 85 + "\n")
        return 0
    else:
        # [X-Ray auto-fix] print(f"\n{RED}{BOLD}✗ Installation Failed - Missing required packages{RESET}")
        # [X-Ray auto-fix] print(f"\nFix with: pip install -r requirements.txt")
        print("=" * 85 + "\n")
        return 1


if __name__ == "__main__":
    sys.exit(main())

File: rust_output/src/verify_install.rs
/// verify_install::py - ZEN_AI_RAG Installation Verification
/// 
/// Checks that all required dependencies are properly installed.
/// Run this after installation to verify everything is ready.
/// 
/// Usage:
/// python verify_install::py

use anyhow::{Result, Context};
use std::collections::HashMap;
use std::path::PathBuf;

pub const GREEN: &str = "\\x1b[92m";

pub const RED: &str = "\\x1b[91m";

pub const YELLOW: &str = "\\x1b[93m";

pub const CYAN: &str = "\\x1b[96m";

pub const RESET: &str = "\\x1b[0m";

pub const BOLD: &str = "\\x1b[1m";

/// Check if a package is installed.
/// 
/// Args:
/// import_name: Name to use in import statement
/// display_name: Human-friendly display name
/// optional: If true, missing is a warning not error
/// 
/// Returns:
/// true if package found, false otherwise
pub fn check_package(import_name: String, display_name: String, optional: bool) -> Result<bool> {
    // Check if a package is installed.
    // 
    // Args:
    // import_name: Name to use in import statement
    // display_name: Human-friendly display name
    // optional: If true, missing is a warning not error
    // 
    // Returns:
    // true if package found, false otherwise
    // try:
    {
        __import__(import_name);
        let mut status = format!("{}✓{}", GREEN, RESET);
        let mut category = if !optional { "REQUIRED".to_string() } else { "OPTIONAL".to_string() };
        println!("  {} {:45} [{}]", status, display_name, category);
        true
    }
    // except ImportError as _e:
}

/// Check if a file/directory exists.
pub fn check_file(filepath: String, display_name: String, optional: bool) -> bool {
    // Check if a file/directory exists.
    let mut path = PathBuf::from(filepath);
    let mut exists = path.exists();
    let mut status = if exists { format!("{}✓{}", GREEN, RESET) } else { format!("{}✗{}", RED, RESET) };
    let mut category = if (optional && !exists) { "WARNING".to_string() } else { if !exists { "REQUIRED".to_string() } else { "OK".to_string() } };
    println!("  {} {:45} [{}]", status, display_name, category);
    exists
}

/// Helper: setup phase for _do_main_setup.
pub fn _do_do_main_setup_setup() -> () {
    // Helper: setup phase for _do_main_setup.
    println!("{}", ("\n".to_string() + ("=".to_string() * 85)));
    println!("{}{}ZEN_AI_RAG Installation Verification{}", BOLD, CYAN, RESET);
    println!("{}", (("=".to_string() * 85) + "\n".to_string()));
    let mut results = HashMap::from([("required".to_string(), HashMap::from([("passed".to_string(), 0), ("failed".to_string(), 0)])), ("optional".to_string(), HashMap::from([("passed".to_string(), 0), ("failed".to_string(), 0)])), ("files".to_string(), HashMap::from([("passed".to_string(), 0), ("failed".to_string(), 0)]))]);
    println!("{}Core Dependencies (REQUIRED):{}", BOLD, RESET);
    let mut core_packages = vec![("nicegui".to_string(), "NiceGUI - Web UI Framework".to_string()), ("uvicorn".to_string(), "Uvicorn - ASGI Server".to_string()), ("requests".to_string(), "Requests - HTTP Client".to_string()), ("httpx".to_string(), "HTTPX - Async HTTP".to_string()), ("bs4".to_string(), "BeautifulSoup4 - HTML Parsing".to_string())];
    for (import_name, display_name) in core_packages.iter() {
        if check_package(import_name, display_name) {
            results["required".to_string()]["passed".to_string()] += 1;
        } else {
            results["required".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    println!("{}LLM & RAG Integration (REQUIRED):{}", BOLD, RESET);
    let mut rag_packages = vec![("qdrant_client".to_string(), "Qdrant Client - Vector Database".to_string()), ("sentence_transformers".to_string(), "Sentence Transformers - Embeddings".to_string()), ("rank_bm25".to_string(), "BM25 - Ranking Algorithm".to_string())];
    for (import_name, display_name) in rag_packages.iter() {
        if check_package(import_name, display_name) {
            results["required".to_string()]["passed".to_string()] += 1;
        } else {
            results["required".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    (display_name, import_name, results)
}

/// Helper: setup phase for main.
pub fn _do_main_setup() -> () {
    // Helper: setup phase for main.
    let (mut display_name, mut import_name, mut results) = _do_do_main_setup_setup();
    println!("{}Document Processing (REQUIRED):{}", BOLD, RESET);
    let mut doc_packages = vec![("PyPDF2".to_string(), "PyPDF2 - PDF Processing".to_string()), ("fitz".to_string(), "PyMuPDF - PDF/Image Processing".to_string()), ("pypdf".to_string(), "PyPDF - Modern PDF Library".to_string())];
    for (import_name, display_name) in doc_packages.iter() {
        if check_package(import_name, display_name) {
            results["required".to_string()]["passed".to_string()] += 1;
        } else {
            results["required".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    println!("{}Audio Processing (REQUIRED for Voice):{}", BOLD, RESET);
    let mut audio_packages = vec![("faster_whisper".to_string(), "Faster Whisper - STT".to_string()), ("piper".to_string(), "Piper TTS - Text-to-Speech".to_string()), ("sounddevice".to_string(), "SoundDevice - Audio I/O".to_string())];
    for (import_name, display_name) in audio_packages.iter() {
        if check_package(import_name, display_name) {
            results["required".to_string()]["passed".to_string()] += 1;
        } else {
            results["required".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    (display_name, import_name, results)
}

/// Helper: setup phase for main.
pub fn _do_main_init() -> () {
    // Helper: setup phase for main.
    let (mut display_name, mut import_name, mut results) = _do_main_setup();
    println!("{}ML & Utilities (REQUIRED):{}", BOLD, RESET);
    let mut util_packages = vec![("numpy".to_string(), "NumPy - Numerical Computing".to_string()), ("scipy".to_string(), "SciPy - Scientific Computing".to_string()), ("pydantic".to_string(), "Pydantic - Data Validation".to_string()), ("dotenv".to_string(), "Python-dotenv - Config Management".to_string())];
    for (import_name, display_name) in util_packages.iter() {
        if check_package(import_name, display_name) {
            results["required".to_string()]["passed".to_string()] += 1;
        } else {
            results["required".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    println!("{}Testing Tools (OPTIONAL):{}", BOLD, RESET);
    let mut test_packages = vec![("pytest".to_string(), "Pytest - Testing Framework".to_string()), ("pytest_asyncio".to_string(), "Pytest AsyncIO - Async Testing".to_string())];
    for (import_name, display_name) in test_packages.iter() {
        if check_package(import_name, display_name, /* optional= */ true) {
            results["optional".to_string()]["passed".to_string()] += 1;
        } else {
            results["optional".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    println!("{}Vision Tools (OPTIONAL):{}", BOLD, RESET);
    let mut vision_packages = vec![("cv2".to_string(), "OpenCV - Computer Vision".to_string()), ("PIL".to_string(), "Pillow - Image Processing".to_string())];
    for (import_name, display_name) in vision_packages.iter() {
        if check_package(import_name, display_name, /* optional= */ true) {
            results["optional".to_string()]["passed".to_string()] += 1;
        } else {
            results["optional".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    (display_name, results)
}

/// Run verification checks.
pub fn main() -> () {
    // Run verification checks.
    let (mut display_name, mut results) = _do_main_init();
    println!("{}Directory Structure:{}", BOLD, RESET);
    let mut dirs = vec![("zena_mode".to_string(), "Core application module".to_string()), ("ui".to_string(), "UI components".to_string()), ("local_llm".to_string(), "Local LLM management".to_string()), ("models".to_string(), "AI model storage directory".to_string()), ("qdrant_storage".to_string(), "Vector database storage".to_string())];
    for (dirname, display_name) in dirs.iter() {
        if check_file(dirname, display_name, /* optional= */ true) {
            results["files".to_string()]["passed".to_string()] += 1;
        } else {
            results["files".to_string()]["failed".to_string()] += 1;
        }
    }
    println!();
    println!("{}", ("=".to_string() * 85));
    println!("{}Summary:{}", BOLD, RESET);
    println!("{}", ("=".to_string() * 85));
    let mut required_ok = results["required".to_string()]["failed".to_string()] == 0;
    results["optional".to_string()]["failed".to_string()] == 0;
    println!("\n  Required Packages: {}{}{} passed, {}{}{} failed", GREEN, results["required".to_string()]["passed".to_string()], RESET, RED, results["required".to_string()]["failed".to_string()], RESET);
    println!("  Optional Packages: {}{}{} passed, {}{}{} warnings", GREEN, results["optional".to_string()]["passed".to_string()], RESET, YELLOW, results["optional".to_string()]["failed".to_string()], RESET);
    println!("  Directory Structure: {}{}{} OK, {}{}{} missing (optional)", GREEN, results["files".to_string()]["passed".to_string()], RESET, YELLOW, results["files".to_string()]["failed".to_string()], RESET);
    if required_ok {
        println!("\n{}{}✓ Installation Verified - Ready to Use!{}", GREEN, BOLD, RESET);
        println!("\nNext: Run 'python zena.py' to start the application");
        println!("{}", (("=".to_string() * 85) + "\n".to_string()));
        0
    } else {
        println!("\n{}{}✗ Installation Failed - Missing required packages{}", RED, BOLD, RESET);
        println!("\nFix with: pip install -r requirements.txt");
        println!("{}", (("=".to_string() * 85) + "\n".to_string()));
        1
    }
}
